File: validateBinarySearchTree.py
import collections 
class TreeNode:
   def __init__(self, x):
     self.val = x
     self.left = None
     self.right = None

class Solution:
  def __init__(self):
    self.isBST = {}

  # @param root, a tree node
  # @return a boolean
  def isValidBST(self, root):
    if not root:
      return False

    stack = []
    stack.append(root)
    while stack:
      n = stack[-1]
      nl = False
      nr = False
      if n.left:
        if not n.left in self.isBST.keys():
          stack.append(n.left)
        else:
          if self.isBST[n.left]:
            nl = True
          else:
            return False
      else:
        nl = True

      if n.right:
        if not n.right in self.isBST.keys():
          stack.append(n.right)
        else:
          if self.isBST[n.right]:
            nr = True
          else:
            return False
      else:
        nr = True

      if nl and nr:
        self.isBST[n] = False
        if ((not n.left) or n.left.val <= n.val) and ((not n.right) or n.val <= n.right.val):
          self.isBST[n] = True

        stack.pop()

    return self.isBST[root]

# test code
t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(0)

t4.left = t2
t4.right = t5
t2.left = t1
t2.right = t3

s = Solution()
print s.isValidBST(t4)

File: validate_binary_search_tree.rb
# Given a binary tree, determine if it is a valid binary search tree (BST).

# Assume a BST is defined as follows:

# The left subtree of a node contains only nodes with keys less than the node's key.
# The right subtree of a node contains only nodes with keys greater than the node's key.
# Both the left and right subtrees must also be binary search trees.

# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Boolean}
def is_valid_bst(root)
  return false if root.nil?
  min_hash = {}
  max_hash = {}

  stack = [root]

  while !stack.empty? do
    node = stack.pop
    return false if min_hash[node] && node.val <= min_hash[node]
    return false if max_hash[node] && node.val >= max_hash[node]

    if node.left
      stack << node.left
      min_hash[node.left] = min_hash[node]
      max_hash[node.left] = node.val
    end

    if node.right
      stack << node.right
      min_hash[node.right] = node.val
      max_hash[node.right] = max_hash[node]
    end
  end

  true
end
